history/history.c

```diff
@@ -0,0 +1,1079 @@
+{
+    int rc = 0, step;
+    sqlite3_stmt *res = NULL;
+    const char *sql_find = "SELECT * FROM rpms WHERE nevra = ?;";
+
+    rc = sqlite3_prepare_v2(db, sql_find, -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_text(res, 1, nevra, -1, NULL);
+    step = sqlite3_step(res);
+    if (step == SQLITE_ROW) {
+        int id = sqlite3_column_int(res, 0);
+        if (pid)
+            *pid = id;
+        sqlite3_finalize(res);
+        /* we found it, return */
+        return 0;
+    }
+    sqlite3_finalize(res); res = NULL;
+
+    /* add it to db */
+    rc = sqlite3_prepare_v2(db, "INSERT INTO rpms(nevra) VALUES (?);", -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_text(res, 1, nevra, -1, NULL);
+    step = sqlite3_step(res);
+    if (step == SQLITE_DONE) {
+        int id = sqlite3_last_insert_rowid(db);
+        if (pid)
+            *pid = id;
+    }
+
+error:
+    if (res)
+        sqlite3_finalize(res);
+    return rc;
+}
+
+/* read transaction items into ht */
+static
+int history_delta_read(sqlite3 *db, struct history_delta *hd, int id)
+{
+    int rc = 0;
+    sqlite3_stmt *res = NULL;
+    const char *sql = "SELECT * FROM trans_items WHERE trans_id = ? ORDER BY rpm_id;";
+    int max_count = 0;
+
+    /* array sizes cannot be larger than number of known rpms */
+    rc = db_rpms_count(db, &max_count);
+    check_db_rc(db, rc);
+
+    hd->added_ids = (int *)calloc(max_count, sizeof(int));
+    check_ptr(hd->added_ids);
+    hd->removed_ids = (int *)calloc(max_count, sizeof(int));
+    check_ptr(hd->removed_ids);
+
+    hd->added_count = 0;
+    hd->removed_count = 0;
+
+    rc = sqlite3_prepare_v2(db, sql, -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_int(res, 1, id);
+    for(int step = sqlite3_step(res); step == SQLITE_ROW; step = sqlite3_step(res)) {
+        int type = sqlite3_column_int(res, 2);
+        int rpm_id = sqlite3_column_int(res, 3);
+        if (type == HISTORY_ITEM_TYPE_ADD || type == HISTORY_ITEM_TYPE_SET) {
+            check_cond(hd->added_count < max_count);
+            hd->added_ids[hd->added_count++] = rpm_id;
+        }
+        else if (type == HISTORY_ITEM_TYPE_REMOVE) {
+            check_cond(hd->removed_count < max_count);
+            hd->removed_ids[hd->removed_count++] = rpm_id;
+        }
+    }
+
+error:
+    if (res)
+        sqlite3_finalize(res);
+    return rc;
+}
+
+/* play one delta transaction on installed_map */
+static
+int db_play_delta(sqlite3 *db, int trans_id,
+                  char *installed_map, int map_size)
+{
+    int rc = 0;
+    sqlite3_stmt *res = NULL;
+    const char *sql = "SELECT * FROM trans_items WHERE trans_id = ? ORDER BY rpm_id;";
+
+    rc = sqlite3_prepare_v2(db, sql, -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_int(res, 1, trans_id);
+    for(int step = sqlite3_step(res); step == SQLITE_ROW; step = sqlite3_step(res)) {
+        int type = sqlite3_column_int(res, 2);
```

Comment:
can we replace 1,2 and 3 with self-explanatory MACRO name

Reply:
I can do that in a later PR. Changing it in this PR will just create too many conflicts for my next change.